### PR TITLE
Instance norm

### PR DESCRIPTION
**Pull request by DmitryUlyanov**
Implements instance normalization modules (1d, 2d, 3d). Based on batch norm, allows affine transform, but  it is switched off by default.

### CODE REVIEW

File: torch/nn/modules/instancenorm.py [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+            input_reshaped, running_mean, running_var, weight, bias,
+            self.training, self.momentum, self.eps)
+
+        # Reshape back
+        self.running_mean.copy_(running_mean.view(b, c).mean(0))
+        self.running_var.copy_(running_var.view(b, c).mean(0))
+
+        return out.view(b, c, *input.size()[2:])
+
+    def force_eval(self):
```
Comment (soumith) [inline torch/nn/modules/instancenorm.py]: is there really value provided with this? I suggest just remove force_eval

File: torch/nn/modules/instancenorm.py [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+        running_mean = self.running_mean.repeat(b)
+        running_var = self.running_var.repeat(b)
+
+        weight, bias = None, None
+        if self.affine:
+            weight = self.weight.repeat(b)
+            bias = self.bias.repeat(b)
+
+        # Apply instance norm
+        input_reshaped = input.view(1, b * c, *input.size()[2:])
```
Comment (soumith) [inline torch/nn/modules/instancenorm.py]: might be worth making `input` contiguous right before this, as view will only work on contiguous inputs:
```
input = input.contiguous()
```

**Review by soumith — CHANGES_REQUESTED** [review]
almost looks good. Please check inline comments.

File: torch/nn/modules/instancenorm.py (hunk 1 above)
Comment (DmitryUlyanov) [inline torch/nn/modules/instancenorm.py]: Removed. Anyway there is a way to force it with `.train(False)`

**Review by soumith — APPROVED** [review]

**Comment by soumith** [comment]
thank you!

**Comment by mys007** [comment]
As `training=False` in fixed in calling `F.batch_norm()`, is it really necessary to:
1) keep running stats at all? One could throw out their keeping as parameters (they could be allocated in `forward() each time anew) as well as repeating them and averaging them back.
2) offer `momentum` in constructor?
